lighthouse-core/gather/computed/dependency-graph/estimator/tcp-connection.js

```diff
@@ -0,0 +1,147 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const INITIAL_CONGESTION_WINDOW = 10;
+const TCP_SEGMENT_SIZE = 1460;
+
+class TcpConnection {
+  /**
+   * @param {number} rtt
+   * @param {number} throughput
+   * @param {number=} responseTime
+   * @param {boolean=} ssl
+   */
+  constructor(rtt, throughput, responseTime = 0, ssl = true) {
+    this._warmed = false;
+    this._ssl = ssl;
+    this._rtt = rtt;
+    this._availableThroughput = throughput;
+    this._responseTime = responseTime;
+    this._congestionWindow = INITIAL_CONGESTION_WINDOW;
+  }
+
+  /**
+   * @param {number} rtt
+   * @param {number} availableThroughput
+   * @return {number}
+   */
+  static maximumSaturatedConnections(rtt, availableThroughput) {
+    const roundTripsPerSecond = 1000 / rtt;
+    const bytesPerRoundTrip = TCP_SEGMENT_SIZE;
+    const bytesPerSecond = roundTripsPerSecond * bytesPerRoundTrip;
+    const minimumThroughputRequiredPerRequest = bytesPerSecond * 8;
+    return Math.floor(availableThroughput / minimumThroughputRequiredPerRequest);
+  }
+
+  /**
+   * @return {number}
+   */
+  _computeMaximumCongestionWindowInSegments() {
+    const bytesPerSecond = this._availableThroughput / 8;
+    const secondsPerRoundTrip = this._rtt / 1000;
+    const bytesPerRoundTrip = bytesPerSecond * secondsPerRoundTrip;
+    return Math.floor(bytesPerRoundTrip / TCP_SEGMENT_SIZE);
+  }
+
+  /**
+   * @param {number} throughput
+   */
+  setThroughput(throughput) {
+    this._availableThroughput = throughput;
+  }
+
+  /**
+   * @param {number} congestion
+   */
+  setCongestionWindow(congestion) {
+    this._congestionWindow = congestion;
+  }
+
+  /**
+   * @param {boolean} warmed
+   */
+  setWarmed(warmed) {
+    this._warmed = warmed;
+  }
+
+  /**
+   * Simulates a network download of a particular number of bytes over an optional maximum amount of time
+   * and returns information about the ending state.
+   *
+   * See https://hpbn.co/building-blocks-of-tcp/#three-way-handshake and
+   *  https://hpbn.co/transport-layer-security-tls/#tls-handshake for details.
+   *
+   * @param {number} bytesToDownload
+   * @param {number=} timeAlreadyElapsed
+   * @param {number=} maximumTimeToElapse
+   * @return {{timeElapsed: number, roundTrips: number, bytesDownloaded: number, congestionWindow: number}}
+   */
+  calculateTimeToDownload(bytesToDownload, timeAlreadyElapsed = 0, maximumTimeToElapse = Infinity) {
```

Comment:
naming threw me... `simulateDownloadingUntil` is way clearer.

kinda weird that  args 1 and 3 are the constraints but not next to eachother. should we just move to an object instead? Seems like we usually would at this point. strawman:

```js
simulateDownloadingUntil({maxBytes, maxMs, msElapsed})
```

Reply:
done, how about switching to an object param in the "incremental PR" after I can rebase against the name change?

Reply:
>  how about switching to an object param in the "incremental PR" after I can rebase against the name change?

sure that's fine.